pkg/util/leaderlease/leaderlease.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// Release tries to delete the leader lock.
+func (e *Etcd) Release() {
+	for i := 0; i < e.renewRetries; i++ {
+		_, err := e.client.CompareAndDelete(e.key, e.value, 0)
+		if err == nil {
+			break
+		}
+		if storage.IsEtcdTestFailed(err) || storage.IsEtcdNotFound(err) {
+			break
+		}
+		util.HandleError(fmt.Errorf("unable to release %s: %v", e.key, err))
+	}
+}
+
+// tryHold attempts to hold on to the lease by repeatedly refreshing its TTL.
+// If the lease hold fails, is deleted, or changed to another user
+func (e *Etcd) tryHold(ttl, index uint64) error {
+	// watch for termination
+	stop := make(chan struct{})
+	lost := make(chan struct{})
+	watchIndex := index
+	go util.Until(func() {
+		if err := e.waitForExpiration(true, &watchIndex, stop); err != nil {
+			util.HandleError(fmt.Errorf("error watching for lease expiration %s: %v", e.key, err))
+			return
+		}
+		glog.V(4).Infof("Lease %s lost due to deletion", e.key)
+		close(lost)
+	}, 100*time.Millisecond, stop)
+	defer close(stop)
+
+	duration := time.Duration(ttl) * time.Second
+	after := time.Duration(float32(duration) * e.waitFraction)
+	last := duration - after
+	interval := last / time.Duration(e.renewRetries)
+	if interval < e.minimumRetryInterval {
+		interval = e.minimumRetryInterval
+	}
+
+	// as long as we can renew the lease, loop
+	for {
+		select {
+		case <-time.After(after):
+			err := wait.Poll(interval, last, func() (bool, error) {
+				glog.V(4).Infof("Renewing lease %s", e.key)
+				resp, err := e.client.CompareAndSwap(e.key, e.value, e.ttl, e.value, index-1)
+				switch {
+				case err == nil:
+					index = eventIndexFor(resp)
+					return true, nil
+				case storage.IsEtcdTestFailed(err):
+					return false, fmt.Errorf("another client has taken the lease %s: %v", e.key, err)
+				case storage.IsEtcdNotFound(err):
+					return false, fmt.Errorf("another client has revoked the lease %s", e.key)
+				default:
+					util.HandleError(fmt.Errorf("unexpected error renewing lease %s: %v", e.key, err))
+					index = etcdIndexFor(err, index)
+					return false, nil
+				}
+			})
+
+			switch err {
+			case nil:
+				// wait again
+				glog.V(4).Infof("Lease %s renewed", e.key)
+			case wait.ErrWaitTimeout:
+				return fmt.Errorf("unable to renew lease %s: %v", e.key, err)
+			default:
+				return fmt.Errorf("lost lease %s: %v", e.key, err)
+			}
+
+		case <-lost:
+			return fmt.Errorf("the lease has been deleted %s", e.key)
+		}
+	}
+}
+
+// waitForExpiration waits until the lease value changes in etcd through deletion, expiration,
+// or explicit change
+func (e *Etcd) waitForExpiration(held bool, from *uint64, stop chan struct{}) error {
+	for {
+		deleted, err := e.waitExpiration(held, from, stop)
+		if err != nil {
+			return err
+		}
+		if deleted {
+			return nil
+		}
+	}
+}
+
+// waitExpiration watches etcd until the lease is deleted, expired, or changed
+func (e *Etcd) waitExpiration(held bool, from *uint64, stop chan struct{}) (bool, error) {
```

Comment:
I'm not a fan of argument mutation.  How muddy does it get to return the values back to callers?  It doesn't look like you take advantage of the fact that changes here might eventually appear in other threads.